src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcServerTransport.cs

```diff
@@ -0,0 +1,201 @@
+        {
+            if (disposing)
+            {
+                _cancellation.Cancel();
+
+                _stream.Dispose();
+
+                _cancellation.Dispose();
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+
+            NamedPipeServerStream connectedStream;
+            try
+            {
+                await _stream.WaitForConnectionAsync(linkedSource.Token).ConfigureAwait(false);
+
+                connectedStream = _stream;
+            }
+            finally
+            {
+                if (!_cancellation.IsCancellationRequested)
+                {
+                    CreateNewPipeServer();
+                }
+            }
+            return connectedStream;
+        }
+
+        private void CreateNewPipeServer()
+        {
+            _stream = new NamedPipeServerStream(
+                _pipeName,
+                PipeDirection.InOut,
+                _maxInstances,
+                PipeTransmissionMode.Byte,
+                PipeOptions.Asynchronous);
+        }
+    }
+
+    internal sealed class UnixDomainSocketServerTransport : IpcServerTransport
+    {
+        private readonly CancellationTokenSource _cancellation = new CancellationTokenSource();
+        private readonly int _backlog;
+        private readonly string _path;
+
+        private UnixDomainSocket _socket;
+
+        public UnixDomainSocketServerTransport(string path, int backlog)
+        {
+            _backlog = backlog;
+            _path = path;
+
+            CreateNewSocketServer();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _cancellation.Cancel();
+
+                try
+                {
+                    _socket.Shutdown(SocketShutdown.Both);
+                }
+                catch { }
+                finally
+                {
+                    _socket.Close(0);
+                }
+                _socket.Dispose();
+
+                _cancellation.Dispose();
+            }
+        }
+
+        public override async Task<Stream> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+            try
+            {
+                Socket socket = await _socket.AcceptAsync(linkedSource.Token).ConfigureAwait(false);
+
+                return new ExposedSocketNetworkStream(socket, ownsSocket: true);
+            }
+            catch (Exception)
+            {
+                // Recreate socket if transport is not disposed.
+                if (!_cancellation.IsCancellationRequested)
+                {
+                    CreateNewSocketServer();
```

Comment:
If the transport wasn't disposed, won't this fail with an `Address already in use` since the file still exists and presumably the socket bound to it as well?

Reply:
I think this largely only fails if the socket is closed. With the fix for the race condition with unlinking and closing you described before, this should be an issue because the file shouldn't exist anymore.

After I refactored the AcceptAsync method that I proposed in https://github.com/dotnet/diagnostics/pull/1303#discussion_r464005190 (not pushed yet, waiting to see if @noahfalk agrees with that change), I'll have to see if this situation is exercised. It might not be because there will be a private Task that is looping over the transport calling AcceptAsync and it is never cancelled until the ReversedDiagnosticsServer is disposed.

Reply:
Because how the C# Socket class calls Dispose from the Close methods, the file should be deleted by the time it gets back to this code path.